### ISSUE TITLE
Commit notifications trigger unexpected link preview

### ISSUE THREAD

**Issue by lieut-data** [mattermost/mattermost-plugin-github#179]
I recently received a notification in the [message of a commit](https://github.com/mattermost/mattermost-sa-code/commit/528462af4dca50e41011fe1076c76edb00327b5b):

![image](https://user-images.githubusercontent.com/1023171/73068914-ca5d3c80-3e82-11ea-9a13-39edad97c9ab.png)

In my summary, this then resulted in a perhaps unexpected link preview compared to all the other links in the message:

![image](https://user-images.githubusercontent.com/1023171/73068947-df39d000-3e82-11ea-8ccb-25bf3b8fd795.png)

**Comment by levb** [comment]
(at a superficial glance) - review and add tests for https://github.com/mattermost/mattermost-plugin-github/blob/23f77f082659e78f1d519eb3aacb0ceb94a3cc21/server/plugin.go#L57?

**Comment by vespian** [comment]
I'll like to take this one.

**Comment by lieut-data** [comment]
Thanks, @vespian!

**Comment by vespian** [comment]
Short update - the problem occurs due to other users doing @mentions of the given user in the commit message. We would need to adjust the regex as pointed out in [2] and adjust the current code [1] to handle rewriting the raw link to the commit (i.e. do not always assume a file/blob) to a nice markdown link.

Will try to prepare a patch.

[1] https://github.com/mattermost/mattermost-plugin-github/blob/ab7832c593851ce7b55882a1325e6e64bf933bc1/server/permalinks.go#L113-L117
[2] https://github.com/mattermost/mattermost-plugin-github/issues/179#issuecomment-578401262

**Comment by Etcaxofjut** [comment]
@lieut-data  

I was playing with the code rewriting the raw link to the commit [4] and realized that I may not have all the data yet.

The `ToDo` text you linked in the Jira description one can trigger directly to bot (`/github todo`) but from what I can see/understand this posts an `ephemeral post`, which in turn does not trigger `MessageWillBePosted` hook (see my question in to ~developers channel [2], I already have enabled `Enable Code Previews` in system console and did some printf-debugging to confirm). 

Another option is while connecting/disconnecting to github[5] the todo list also gets displayed, but I guess this is not the case here. 

The third option is the API call via ServeHTTP-`/api/v1/todo`, but I can't find it being called anywhere.

I wonder if I misconfigured somehow my dev instance. How can I recreate it reliably?

Also, if by chance you could help me one more thing - the `MessageWillBePosted` hook is called with the userID of `github` bot, which does not seem to have GH credentials stored and results in error [3]. I suppose that I did something wrong while setting up my test instance which results in the userID of the bot instead of that of the user making the post being called. Do you have any ideas maybe?

Thanks in advance for any information.


[1] [issue-mattermost-plugin-github-179](https://github.com/mattermost/mattermost-plugin-github/issues/179)
[2] [<jump to convo>](/core/pl/n8rzet9gfjgyfr9kgcafh7xnfy)
[3] https://github.com/mattermost/mattermost-plugin-github/blob/6e7a3ef8df706cbaf39c579e42d6452b0cbfe968/server/plugin.go#L222
[4] https://github.com/mattermost/mattermost-plugin-github/blob/ab7832c593851ce7b55882a1325e6e64bf933bc1/server/permalinks.go#L113-L117
[5] https://github.com/mattermost/mattermost-plugin-github/blob/6e7a3ef8df706cbaf39c579e42d6452b0cbfe968/server/api.go#L292

**Comment by vespian** [comment]
@lieut-data

I was playing with the code rewriting the raw link to the commit [4] and realized that I may not have all the data yet.

The ToDo text you linked in the Jira description one can trigger directly to bot (`/github todo`) but from what I can see/understand this posts an ephemeral post, which in turn does not trigger MessageWillBePosted hook (see my question in to ~developers channel [2], I already have enabled Enable Code Previews in system console and did some printf-debugging to confirm).

Another option is while connecting/disconnecting to github[5] the todo list also gets displayed, but I guess this is not the case here.

The third option is the API call via `ServeHTTP-/api/v1/todo`, but I can't find it being called anywhere.

I wonder if I misconfigured somehow my dev instance. How can I recreate it reliably?

Also, if by chance you could help me one more thing - the `MessageWillBePosted` hook is called with the userID of github bot, which does not seem to have GH credentials stored and results in error [3]. I suppose that I did something wrong while setting up my test instance which results in the userID of the bot instead of that of the user making the post being called. Do you have any ideas maybe?

Thanks in advance for any information.

[1] issue-mattermost-plugin-github-179
[2] https://github.com/mattermost/mattermost-plugin-github/blob/ab7832c593851ce7b55882a1325e6e64bf933bc1/server/permalinks.go#L113-L117
[3] https://github.com/mattermost/mattermost-plugin-github/blob/6e7a3ef8df706cbaf39c579e42d6452b0cbfe968/server/plugin.go#L222
[4] https://github.com/mattermost/mattermost-plugin-github/blob/ab7832c593851ce7b55882a1325e6e64bf933bc1/server/permalinks.go#L113-L117
[5] https://github.com/mattermost/mattermost-plugin-github/blob/6e7a3ef8df706cbaf39c579e42d6452b0cbfe968/server/api.go#L292

**Comment by vespian** [comment]
Also,  the todo list output looks like this for me:
![todo](https://user-images.githubusercontent.com/2124609/74103120-b9cae880-4b49-11ea-8310-35d211b63d6a.png)

**Comment by lieut-data** [comment]
@vespian, `MessageWillBePosted` will indeed only be called for non-ephemeral posts, but in this particular case, I get a regular post from the GitHub plugin's morning reminder. You can enable this with `/github settings reminders on`.

re: the GitHub bot and `UserId`, I might be misunderstanding, but it is indeed the GitHub bot making these posts, so that part seems correct. Does the code preview feature rely on the posting user's GitHub credentials to make the underlying API request? That might be a problem if users who don't link their GitHub accounts post such messages. I don't recall if this is an existing problem with the plugin, but perhaps this particular request should be made with the "default" GitHub credentials?

**Comment by vespian** [comment]
RE: Github credentials - I have prepared a simple patch:

https://github.com/mattermost/mattermost-plugin-github/pull/191

It does not solve the main issue though, just fixes a thing that most likelly is a different bug. Please check the PR description for details.

**Comment by levb** [comment]
I am so very confused (and apologies for my ignorance about how the plugin works).

This is a message that is generated by the plugin itself, do we really use MessageWillBePosted to rewrite it? Seems illogical, why not compose correctly to start with?

Also, FWIW, if we _are_ setting up regex link rewrites from the GitHub plugin, perhaps we should be promoting the autolink plugin, and programmatically set it up from GitHub, like we are doing with Jira now?

cc @lieut-data @crspeller @jfrerich

**Comment by vespian** [comment]
@lieut-data 

Please bear with me, I still had no luck in recreating the error, even with #191. Is it possible that the previews you mentioned are done through autolinks plugin?

**Comment by lieut-data** [comment]
@vespian, ahh, perhaps it is autolink that's at play here, and not GitHub at all. That makes a lot more sense: and thank you for digging to the bottom of this.

@levb, would it makes sense to guard autolink with [ShouldProcessMessage](https://godoc.org/github.com/mattermost/mattermost-server/plugin#HelpersImpl.ShouldProcessMessage)?

**Comment by levb** [comment]
> would it makes sense to guard autolink with `ShouldProcessMessage`?

@lieut-data Yes, I think so. Do you think we should expose the options in the Config? Do we have common code to do that?

Separately, I also think we should work on removing link substitution from GitHub and delegate it all to Autolink, programmatically, like Jira does now. This would be better for performance, and promote plugin usage. cc @aaronrothschild @jfrerich.

Now, what do we do with this issue then, close it?

The relevant Autolink config on community is 
![image](https://user-images.githubusercontent.com/1187448/74667610-6f082c80-5158-11ea-8492-f201673bc684.png)

**Comment by lieut-data** [comment]
@levb, I'd be happy to move this issue over to the autolink repository to keep the history. Agreed re: investigating programmatically managing link substitution, but as a separate issue?
